supabase/migrations/20230817110845_access_token.sql

```diff
@@ -0,0 +1,174 @@
+    token_name text
+)
+    returns text
+    language plpgsql
+    strict
+as $$
+declare
+    user_id uuid = auth.uid();
+    account app.accounts = account from app.accounts account where id = user_id;
+    token bytea = gen_random_bytes(16);
+    token_hash bytea = pgsodium.crypto_pwhash_str(token);
+    token_text text = encode(token, 'hex');
+begin
+    begin
+        insert into app.access_tokens(user_id, token_hash, token_name)
+        values (account.id, token_hash, token_name);
+    exception when unique_violation then
+        raise exception 'Token with name `%s` already exists', token_name;
+    end;
+
+    return replace(user_id::text, '-', '') || token_text;
+end;
+$$;
+
+create type app.access_token_struct as (
+    id uuid,
+    token_name text,
+    created_at timestamptz
+);
+
+create or replace function public.get_access_tokens()
+    returns setof app.access_token_struct
+    language plpgsql
+    strict
+as $$
+declare
+    user_id uuid = auth.uid();
+    account app.accounts = account from app.accounts account where id = user_id;
+    token bytea = gen_random_bytes(16);
+    token_hash bytea = pgsodium.crypto_pwhash_str(token);
+    token_text text = encode(token, 'hex');
+begin
+    return query
+    select id, token_name, created_at
+    from app.access_tokens at
+    where at.user_id = account.id;
+end;
+$$;
+
+create or replace function public.delete_access_token(
+    token_id uuid
+)
+    returns void
+    language plpgsql
+    strict
+as $$
+declare
+    user_id uuid = auth.uid();
+    account app.accounts = account from app.accounts account where id = user_id;
+begin
+    delete from app.access_tokens at
+    where at.user_id = account.id and at.id = token_id;
+end;
+$$;
+
+create or replace function public.redeem_access_token(
+    access_token text
+)
+    returns text
+    language plpgsql
+    security definer
+    strict
+as $$
+declare
+    user_id_text text;
+    user_id uuid;
+    token_text text;
+    token bytea;
+    access_token_record record;
+    token_hash bytea;
+    token_valid boolean;
+    now timestamp;
+    one_hour_from_now timestamp;
+    issued_at int;
+    expiry_at int;
+    jwt_secret text;
+begin
+    -- validate access token
+    if length(access_token) != 64 then
+        raise exception 'Invalid token';
+    end if;
+
+    user_id_text := substring(access_token from 1 for 32);
+    user_id := substring(user_id_text from 1 for 8) || '-' ||
+                   substring(user_id_text from 9 for 4) || '-' ||
+                   substring(user_id_text from 13 for 4) || '-' ||
+                   substring(user_id_text from 17 for 4) || '-' ||
+                   substring(user_id_text from 21 for 12);
+    token_text := substring(access_token from 33 for 32);
+    token := decode(token_text, 'hex');
+
+    for token_hash in
+        select t.token_hash from app.access_tokens t where t.user_id = user_id
+    loop
+        token_valid := pgsodium.crypto_pwhash_str_verify(token_hash, token);
```

Comment:
Again, no need to do password hash verification, which is also VERY CPU intensive. When you switch to SHA256, make sure you don't just do `=` as that's not constant-time comparison. I need to look into how this can be achieved in Postgres easily as I'm not finding it but @michelp this may be a good API to add to pgsodium.

In general though, this is really a fringe security benefit for this use case.

Reply:
Oh also I suppose libsodium uses Argon2 which is _incredibly_ CPU and memory intensive (depending on the settings). So this is really an easy way to crash the database -- just start generating / verifying tokens all the time.

Reply:
Moved to a SHA256 based impl.